Sources/AardvarkReveal/RevealAttachmentGenerator.swift

```diff
@@ -0,0 +1,518 @@
+        mainScreenParameters.completion = { [delegate] in
+            DispatchQueue.main.async {
+                delegate?.revealAttachmentGeneratorDidCaptureMainScreenSnapshot()
+            }
+        }
+        taskParameters.append(mainScreenParameters)
+
+        let identifiers = applicationState.application.identifiersForObjectsInHierachyWithImages()
+        for identifier in identifiers {
+            taskParameters.append(try resourceImageParameters(for: identifier, showingSubviews: false))
+            taskParameters.append(try resourceImageParameters(for: identifier, showingSubviews: true))
+        }
+
+        return taskParameters
+    }
+
+    private func fetchImages(
+        using taskParameters: [DownloadTaskParameters],
+        urlSession: URLSession,
+        archiveBuilder: ARKRevealArchiveBuilder,
+        completionQueue: DispatchQueue,
+        completion: @escaping () -> Void
+    ) {
+        let dispatchGroup = DispatchGroup()
+
+        for parameters in taskParameters {
+            dispatchGroup.enter()
+            let task = urlSession.dataTask(with: parameters.urlRequest) { data, _, _ in
+                // Try to add the data to the archive if we got valid data back. The Reveal file is still valid even if
+                // it's missing some images, so don't throw any errors if this fails.
+                if let data = data {
+                    try? archiveBuilder.addFile(atPath: parameters.pathInArchive, with: data)
+                }
+
+                parameters.completion?()
+
+                dispatchGroup.leave()
+            }
+            task.resume()
+        }
+
+        dispatchGroup.notify(queue: completionQueue) {
+            completion()
+        }
+    }
+
+    private func applicationName() -> String {
+        return Bundle.main.localizedInfoDictionary?["CFBundleDisplayName"] as? String
+            ?? Bundle.main.localizedInfoDictionary?["CFBundleName"] as? String
+            ?? Bundle.main.infoDictionary?["CFBundleDisplayName"] as? String
+            ?? Bundle.main.infoDictionary?["CFBundleName"] as? String
+            ?? "Unknown App"
+    }
+
+    private func addPropertiesPlist(appName: String, bundleName: String, archiveBuilder: ARKRevealArchiveBuilder) throws {
+        let propertiesPlist: [String: Any] = [
+            "application-name": appName,
+            "version": 2,
+        ]
+        let propertiesFileURL = URL(
+            fileURLWithPath: (NSTemporaryDirectory() as NSString).appendingPathComponent("\(appName)-Properties.plist")
+        )
+        try (propertiesPlist as NSDictionary).write(to: propertiesFileURL)
+        try archiveBuilder.addFile(
+            atPath: "\(bundleName)/Properties.plist",
+            with: try Data(contentsOf: propertiesFileURL)
+        )
+        try? FileManager.default.removeItem(at: propertiesFileURL)
+    }
+
+}
```

Comment:
I'm struggling with how this class can be modified confidently in the future. Can we add some unit or contractural tests?

Reply:
- [ ] Add tests for attachment generator

Reply:
Added some integration tests for the generator's callbacks in different scenarios. Still want to get one more big one in to test its interaction with the archive builder.

Reply:
Pretty happy with the test setup I've got now. Probably a few more edge cases that could be covered, but I think it's got pretty good coverage already.